src/thor/route_action.cc

```diff
@@ -162,50 +262,69 @@ std::list<valhalla::odin::TripPath> thor_worker_t::path_arrive_by(
     thor::PathAlgorithm* path_algorithm = get_path_algorithm(costing, *origin, *destination);
     path_algorithm->Clear();
 
+    // TODO: delete this and send all cases to the function above
     // If we are continuing through a location we need to make sure we
     // only allow the edge that was used previously (avoid u-turns)
-    while (!path.empty() && destination->path_edges_size() > 1) {
-      if (destination->path_edges().rbegin()->graph_id() == path.front().edgeid) {
+    bool through = destination->type() == odin::Location::kThrough ||
+                   destination->type() == odin::Location::kBreakThrough;
+    while (through && first_edge.Is_Valid() && destination->path_edges_size() > 1) {
+      if (destination->path_edges().rbegin()->graph_id() == first_edge) {
         destination->mutable_path_edges()->SwapElements(0, destination->path_edges_size() - 1);
       }
       destination->mutable_path_edges()->RemoveLast();
     }
 
     // Get best path and keep it
     auto temp_path = get_path(path_algorithm, *origin, *destination, costing);
-    temp_path.swap(path);
+    first_edge = temp_path.front().edgeid;
+    temp_path.swap(path); // so we can append to path instead of prepend
 
     // Merge through legs by updating the time and splicing the lists
     if (!temp_path.empty()) {
       auto offset = path.back().elapsed_time;
       std::for_each(temp_path.begin(), temp_path.end(),
                     [offset](PathInfo& i) { i.elapsed_time += offset; });
+      // Connects via the same edge so we only need it once
       if (path.back().edgeid == temp_path.front().edgeid) {
         path.pop_back();
+      } else if (destination->type() == odin::Location::kVia) {
+        // Insert a route discontinuity if the paths meet at opposing edges and not
+        // at a graph node. Use path size - 1 as the index where the discontinuity lies.
+        via_discontinuity(*reader, *destination, path.back().edgeid, temp_path.front().edgeid, vias,
+                          temp_path.size(), true);
```

Comment:
here we detect the case where we need to trim shape in the middle of an edge. only `via` types will have this as only they will even allow you to make a uturn in the middle of an edge (whout arriving)